jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java

```diff
@@ -0,0 +1,210 @@
+ * details about the dp algorithm.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Alexandru Valeanu
+ */
+public class HeldKarpTSP<V, E>
+        implements TSPAlgorithm<V, E> {
+
+    /**
+     * Construct a new instance
+     */
+    public HeldKarpTSP() {
+    }
+
+    private void printState(int state, int n){
+        System.out.printf("State %d: ", state);
+        for (int i = 0; i < n; i++) {
+            if (((state >> i) & 1) > 0){
+                System.out.printf("%d ", i);
+            }
+        }
+        System.out.println();
+    }
+
+    private double memo(int previousNode, int state, double[][] C, double[][] W){
+        // have we seen this state before?
+        if (C[previousNode][state] != Double.MIN_VALUE)
+            return C[previousNode][state];
+
+        // no cycle has been found yet
+        double totalCost = Double.MAX_VALUE;
+
+        // check if all nodes have been visited (i.e. state + 1 == 2^n)
+        if (state == (1 << W.length) - 1){
+            // check if there is a return edge we can use
+            if (W[previousNode][0] != Double.MAX_VALUE)
+                totalCost = W[previousNode][0];
+            else
+                totalCost = Double.MAX_VALUE;
+        }
+        else{
+            // try to find the 'best' next (i.e. unvisited and adjacent to previousNode) node in the tour
+            for (int i = 0; i < W.length; i++) {
+                if (((state >> i) & 1) == 0 && W[previousNode][i] != Double.MAX_VALUE){
+                    totalCost = Math.min(totalCost, W[previousNode][i] + memo(i, state ^ (1 << i), C, W));
+                }
+            }
+        }
+
+        return C[previousNode][state] = totalCost;
+    }
+
+    /**
+     * Computes a minimum-cost Hamiltonian tour.
+     *
+     * @param graph the input graph
+     * @return a minimum-cost tour if one exists, null otherwise
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    @Override
+    public GraphPath<V, E> getTour(Graph<V, E> graph) {
+        final int n = graph.vertexSet().size(); // number of nodes
+
+        if (n == 0) {
+            throw new IllegalArgumentException("Graph contains no vertices");
+        }
+
+        if (n == 1){
+            V startNode = graph.vertexSet().iterator().next();
+            return new GraphWalk<>(graph, startNode, startNode, Collections.singletonList(startNode), null, 0);
+        }
+
+        // W[u, v] = the cost of the minimum weight between u and v
+        double[][] W = new double[n][n];
+        for (int i = 0; i < n; i++) {
+            Arrays.fill(W[i], Double.MAX_VALUE);
+        }
+
+        /*
+         * Normalize the graph
+         *   map each vertex to an integer (using a HashMap)
+         *   keep the reverse mapping  (using an ArrayList)
+         */
+        Map<V, Integer> mapToInt = new HashMap<>();
```

Comment:
In several other implementations, the mapToInt and mapToV are called respectively:
vertexMap and indexList. Consider naming these the same.